src/main/java/com/google/firebase/internal/FirebaseExecutors.java

```diff
@@ -16,20 +16,122 @@
+
+import java.util.HashSet;
+import java.util.Set;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
 
-/** Default executors used for internal Firebase threads. */
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Default ThreadManager implementations used by the Admin SDK. */
 public class FirebaseExecutors {
 
-  public static final ScheduledExecutorService DEFAULT_SCHEDULED_EXECUTOR;
+  private static final Logger logger = LoggerFactory.getLogger(FirebaseExecutors.class);
+
+  public static final ThreadManager DEFAULT_THREAD_MANAGER;
 
   static {
     if (GaeThreadFactory.isAvailable()) {
-      DEFAULT_SCHEDULED_EXECUTOR = GaeThreadFactory.DEFAULT_EXECUTOR;
+      DEFAULT_THREAD_MANAGER = new GaeThreadManager();
     } else {
-      DEFAULT_SCHEDULED_EXECUTOR =
-          Executors.newSingleThreadScheduledExecutor(Executors.defaultThreadFactory());
+      DEFAULT_THREAD_MANAGER = new DefaultThreadManager();
+    }
+  }
+
+  /**
+   * An abstract ThreadManager implementation that uses the same executor service
+   * across all active apps. The executor service is initialized when the first app is initialized,
+   * and terminated when the last app is deleted. This class is thread safe.
+   */
+  abstract static class GlobalThreadManager extends ThreadManager {
+
+    private final Set<String> apps = new HashSet<>();
+    private ScheduledExecutorService executorService;
+
+    @Override
+    protected synchronized ScheduledExecutorService getExecutor(FirebaseApp app) {
+      if (executorService == null) {
+        executorService = doInit();
+      }
+      apps.add(app.getName());
+      return executorService;
+    }
+
+    @Override
+    protected synchronized void releaseExecutor(
+        FirebaseApp app, ScheduledExecutorService executor) {
+      if (apps.remove(app.getName()) && apps.isEmpty()) {
+        doCleanup(executorService);
+        executorService = null;
+      }
+    }
+
+    /**
+     * Initializes the executor service. Called when the first application is initialized.
+     */
+    protected abstract ScheduledExecutorService doInit();
+
+    /**
+     * Cleans up the executor service. Called when the last application is deleted.
+     */
+    protected abstract void doCleanup(ScheduledExecutorService executorService);
+  }
+
+  private static class DefaultThreadManager extends GlobalThreadManager {
+
+    @Override
+    protected ScheduledExecutorService doInit() {
+      int cores = Runtime.getRuntime().availableProcessors();
+      // Create threads as daemons to ensure JVM exit when all foreground jobs are complete.
+      ThreadFactory threadFactory = new ThreadFactoryBuilder()
+          .setNameFormat("firebase-default-%d")
+          .setDaemon(true)
+          .setThreadFactory(getThreadFactory())
+          .build();
+      logger.debug("Initializing default executor with {} max threads", cores);
```

Comment:
I don't believe there is any guarantee that ScheduledThreadPoolExecutor will try to schedule each thread on a dedicated core, so using this number here may not make much sense.

Reply:
Correct. I just used that as a max limit since `ScheduledThreadPoolExecutor` must be given some core thread pool count. I think we can go back to cached thread pool when we implement #77.